### PR TITLE
feat: dataset management — full feature with tests

### PR DESCRIPTION
**Pull request by jariy17**
## Summary

Implements complete dataset management lifecycle for AgentCore CLI:

- **Schema & Primitives**: DatasetPrimitive (add/remove/preview) with PREDEFINED_V1 and SIMULATED_V1 schema types, name validation, kmsKeyArn support
- **CLI Commands**: `add dataset`, `remove dataset`, `dataset download`, `dataset publish-version`, `dataset remove-version`
- **Operations**: Incremental push engine (diff by exampleId, batched API calls with retry), post-deploy sync (hash-based change detection), dataset-driven evaluation (load scenarios, invoke agent, collect spans)
- **TUI**: Add dataset wizard, dataset hub (download/publish/remove-version), dataset picker in batch eval flow
- **Bug Fixes**: Deploy preflight now recognizes datasets as deployable resources; removed stale x-amz-expected-bucket-owner header workaround

## Test plan

- [x] 68 unit tests (DatasetPrimitive, AWS client, push engine, pull, publish, resolve, wait, post-deploy sync, session provider, preflight)
- [x] 17 integration tests (undeployed commands, validation edge cases, scaffolding verification, file preservation)
- [x] 19 E2E tests (full lifecycle, 1000-example large batch, dataset-driven eval with Builtin.Faithfulness)
- [x] All E2E tests verified in ap-southeast-2 against account 619071331382

### CODE REVIEW

File: src/cli/operations/deploy/post-deploy-datasets.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      const pushResult = await pushDataset({
+        region,
+        datasetId: state.datasetId,
+        localFilePath,
+        configBaseDir,
+      });
+
+      updatedDatasets[dataset.name] = {
+        ...state,
+        contentHash: currentHash,
```
Comment (tejaskash) [inline src/cli/operations/deploy/post-deploy-datasets.ts]: Stale hash: `currentHash` is computed before `pushDataset` runs, but push rewrites the file with new `exampleId`s. So the saved hash never matches the on-disk file, and every subsequent `deploy` will re-run `pushDataset` (no-op diff, but still a network round-trip per dataset).

Fix: re-read the file after push and hash that, or have `pushDataset` return the final content/hash.

File: src/cli/operations/eval/shared/span-collector.ts [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+export async function collectSpans(options: CollectSpansOptions): Promise<CollectedSpans> {
+  const { sessionIds, querySpans, onProgress } = options;
+
+  // Phase 1: Wait for CloudWatch ingestion
+  onProgress?.(0, sessionIds.length, `Waiting for span ingestion (${SPAN_INGESTION_DELAY_MS / 1000}s)...`);
+  await sleep(SPAN_INGESTION_DELAY_MS);
+
+  // Phase 2: Poll each session in parallel
+  let collectedCount = 0;
+  const results = await Promise.all(
```
Comment (tejaskash) [inline src/cli/operations/eval/shared/span-collector.ts]: `Promise.all` rejects on the first throw — if one session hits `AccessDenied`, every other in-flight poll gets aborted. `Promise.allSettled` would let one bad session fail without taking the whole eval down.

File: src/cli/operations/dataset/publish.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ */
+export async function publishDataset(options: PublishOptions): Promise<PublishResult> {
+  const { region, datasetId } = options;
+
+  const versionResult = await createDatasetVersion({ region, datasetId });
+
+  await waitForDatasetActive(region, datasetId);
+
+  // Re-fetch to get final state after publish
+  const { getDataset } = await import('../../aws/agentcore-datasets');
```
Comment (tejaskash) [inline src/cli/operations/dataset/publish.ts]: Why the dynamic import? `createDatasetVersion` is already statically imported from this same module on line 4 — no cycle to dodge. Move `getDataset` up there.

**Review by tejaskash — COMMENTED** [review]
Nice work overall — clean layering, good test coverage. A few inline notes; only the post-deploy hash one feels like it should block.

File: src/cli/aws/agentcore-datasets.ts [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+  nextToken?: string;
+}
+
+// ============================================================================
+// HTTP signing helper
+// ============================================================================
+
+function getControlPlaneEndpoint(region: string): string {
+  const stage = process.env.AGENTCORE_STAGE?.toLowerCase();
+  if (stage === 'beta') return `https://beta.${region}.elcapcp.genesis-primitives.aws.dev`;
```
Comment (tejaskash) [inline src/cli/aws/agentcore-datasets.ts:151]: Just flagging — these `aws.dev` hosts are Amazon-internal and will ship as string literals in the public tarball. Gated by `AGENTCORE_STAGE` so only matters if someone sets it, but worth confirming this is how other primitives handle staging (vs. e.g. routing through the SDK's endpoint resolver).

File: src/cli/aws/agentcore-datasets.ts [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+  if (stage === 'gamma') return `https://gamma.${region}.elcapcp.genesis-primitives.aws.dev`;
+  return `https://bedrock-agentcore-control.${region}.${dnsSuffix(region)}`;
+}
+
+async function signedRequest(options: {
```
Comment (tejaskash) [inline src/cli/aws/agentcore-datasets.ts:156]: No timeout on `fetch()`. A hung connection blocks indefinitely and your retry never fires. `AbortSignal.timeout(...)` would help.

File: src/cli/primitives/DatasetPrimitive.ts [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  addScreen(): AddScreenComponent {
+    return null;
+  }
+
+  /**
+   * Copy the starter JSONL asset file to the dataset location.
+   */
+  private async scaffoldDatasetFile(name: string, schemaType: string, location: string): Promise<void> {
```
Comment (tejaskash) [inline src/cli/primitives/DatasetPrimitive.ts:218]: nit: `name` param is unused.

**Review by tejaskash — APPROVED** [review]
Verified the four blocking/sub-blocking comments are addressed:

- post-deploy-datasets.ts: now re-reads the file after `pushDataset` and hashes the post-push content ✓
- span-collector.ts: switched to `Promise.allSettled` so one failure doesn't abort the rest ✓
- publish.ts: dynamic `getDataset` import promoted to top-level ✓
- agentcore-datasets.ts: `fetch()` now uses `AbortSignal.timeout(30_000)` ✓

LGTM.

File: src/schema/schemas/primitives/dataset.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  schemaType: DatasetSchemaTypeSchema,
+  /** Optional description (max 200 characters) */
+  description: z.string().max(200).optional(),
+  /** Dataset content management config */
+  config: DatasetConfigSchema,
+  /** Optional KMS key ARN for SSE-KMS encryption. Immutable after creation. */
+  kmsKeyArn: z
+    .string()
+    .regex(/^arn:aws(-[a-z]+)*:kms:[a-zA-Z0-9-]*:[0-9]{12}:key\/[a-zA-Z0-9-]{36}$/, 'Must be a valid KMS key ARN')
+    .optional(),
```
Comment (padmak30) [inline src/schema/schemas/primitives/dataset.ts:64]: Is the kms key wired through in the CLI flow?

**Review by padmak30 — COMMENTED** [review]
Some of the back navigation is broken.
1. From run -> batch evals -> eval source (dataset & version) -> evaluators and navigating back with escape takes me back to run skipping the eval source step.
2. Back-navigation in AddDatasetScreen loses previously typed name.
3. DatasetFlow back-nav resets datasets to [], leaving the hub permanently stuck with "No items available" after pressing Esc from any sub-screen.

File: src/schema/schemas/primitives/dataset.ts (hunk 7 above)
Comment (jariy17) [reply src/schema/schemas/primitives/dataset.ts:64]: It should've been but when I rebased my changes on public main, it might dropped the flag. This will be a fast follow up along with the tui
